File: CS/CustomPrintPreview/Grid/CustomGridView.cs
using System;
using System.Drawing;
using DevExpress.XtraGrid;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Base;

namespace CustomPrintPreview.Grid {
    public class CustomGridView :GridView {
        public CustomGridView(GridControl ownerGrid): base(ownerGrid) { }
        public CustomGridView() { }

        internal const string CustomName = "CustomGridView";
        protected override string ViewName {
            get { return CustomName; }
        }

        protected override UserControl PrintDesigner {
            get {
                UserControl ctrl = new UserControl();
                GetGridPreviewEventArgs args = GetGridPreview();
                CustomGridViewPrinting gvp = new CustomGridViewPrinting(args.GridPreview,
                    args.GroupColumn);
                args.GridPreview = null;
                args.GroupColumn = null;
                gvp.InitFrame(this, "GridView", new Bitmap(16, 16));
                gvp.Dock = DockStyle.Fill;
                ctrl.Controls.Add(gvp);
                ctrl.Visible = true;
                ctrl.Dock = DockStyle.Fill;
                ctrl.Size = gvp.UserControlSize;
                return ctrl;
            }
        }

        GetGridPreviewEventArgs GetGridPreview() {
            EventHandler<GetGridPreviewEventArgs> handler = Events[fGetGridPreview]
                as EventHandler<GetGridPreviewEventArgs>;
            if (handler == null) return null;
            GetGridPreviewEventArgs args = new GetGridPreviewEventArgs();
            handler(this, args);
            return args;
        }

        public override void Assign(BaseView v, bool copyEvents) {
            BeginUpdate();
            try {
                base.Assign(v, copyEvents);
                if (copyEvents)
                    GetCustomGridPreview += (EventHandler<GetGridPreviewEventArgs>)((CustomGridView)v).Events[fGetGridPreview];
            } finally { EndUpdate(); }
        }

        internal void FireChangedInternal() {
            base.FireChanged();
        }

        static readonly object fGetGridPreview = new object();
        public event EventHandler<GetGridPreviewEventArgs> GetCustomGridPreview {
            add { Events.AddHandler(fGetGridPreview, value); }
            remove { Events.RemoveHandler(fGetGridPreview, value); }
        }
    }
}

File: VB/CustomPrintPreview/Grid/CustomGridView.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports DevExpress.XtraGrid
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Base

Namespace CustomPrintPreview.Grid
	Public Class CustomGridView
		Inherits GridView
		Public Sub New(ByVal ownerGrid As GridControl)
			MyBase.New(ownerGrid)
		End Sub
		Public Sub New()
		End Sub

		Friend Const CustomName As String = "CustomGridView"
		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return CustomName
			End Get
		End Property

		Protected Overrides ReadOnly Property PrintDesigner() As UserControl
			Get
				Dim ctrl As New UserControl()
				Dim args As GetGridPreviewEventArgs = GetGridPreview()
				Dim gvp As New CustomGridViewPrinting(args.GridPreview, args.GroupColumn)
				args.GridPreview = Nothing
				args.GroupColumn = Nothing
				gvp.InitFrame(Me, "GridView", New Bitmap(16, 16))
				gvp.Dock = DockStyle.Fill
				ctrl.Controls.Add(gvp)
				ctrl.Visible = True
				ctrl.Dock = DockStyle.Fill
				ctrl.Size = gvp.UserControlSize
				Return ctrl
			End Get
		End Property

		Private Function GetGridPreview() As GetGridPreviewEventArgs
			Dim handler As EventHandler(Of GetGridPreviewEventArgs) = TryCast(Events(fGetGridPreview), EventHandler(Of GetGridPreviewEventArgs))
			If handler Is Nothing Then
				Return Nothing
			End If
			Dim args As New GetGridPreviewEventArgs()
			handler(Me, args)
			Return args
		End Function

		Public Overrides Sub Assign(ByVal v As BaseView, ByVal copyEvents As Boolean)
			BeginUpdate()
			Try
				MyBase.Assign(v, copyEvents)
				If copyEvents Then
                    AddHandler GetCustomGridPreview, CType(Events(fGetGridPreview), EventHandler(Of GetGridPreviewEventArgs))
				End If
			Finally
				EndUpdate()
			End Try
		End Sub

		Friend Sub FireChangedInternal()
			MyBase.FireChanged()
		End Sub

		Private Shared ReadOnly fGetGridPreview As Object = New Object()
		Public Custom Event GetCustomGridPreview As EventHandler(Of GetGridPreviewEventArgs)
			AddHandler(ByVal value As EventHandler(Of GetGridPreviewEventArgs))
				Events.AddHandler(fGetGridPreview, value)
			End AddHandler
			RemoveHandler(ByVal value As EventHandler(Of GetGridPreviewEventArgs))
				Events.RemoveHandler(fGetGridPreview, value)
			End RemoveHandler
            RaiseEvent(ByVal sender As System.Object, ByVal e As GetGridPreviewEventArgs)
            End RaiseEvent
		End Event
	End Class
End Namespace
